packages/smart-contracts/src/contracts/ERC20SwapToPay.sol

```diff
@@ -0,0 +1,139 @@
+    erc20.approve(address(swapRouter), max);
+  }
+  
+  /**
+  * @notice Performs a token swap between a payment currency and a request currency, and then
+  *         calls a payment proxy to pay the request, including fees.
+  * @param _to Transfer recipient = request issuer
+  * @param _amount Amount to transfer in request currency
+  * @param _amountInMax Maximum amount allowed to spend for currency swap, in payment currency.
+            This amount should take into account the fees.
+    @param _path, path of ERC20 tokens to swap from requestedToken to spentToken. The first 
+            address of the path should be the payment currency. The last element should be the
+            request currency.
+  * @param _paymentReference Reference of the payment related
+  * @param _feeAmount Amount of the fee in request currency
+  * @param _feeAddress Where to pay the fee
+  * @param _deadline Deadline for the swap to be valid
+  */
+  function swapTransferWithReference(
+    address _to,
+    uint256 _amount,      // requestedToken
+    uint256 _amountInMax, // spentToken
+    address[] calldata _path, // from requestedToken to spentToken
+    bytes calldata _paymentReference,
+    uint256 _feeAmount,   // requestedToken
+    address _feeAddress,
+    uint256 _deadline
+  )
+    external
+  {
+    ERC20 spentToken = ERC20(_path[0]);
+    ERC20 requestedToken = ERC20(_path[_path.length-1]);
+    
+    uint256 requestedTotalAmount = _amount + _feeAmount;
+
+    require(spentToken.allowance(msg.sender, address(this)) > _amountInMax, "Not sufficient allowance for swap to pay.");
+    require(spentToken.safeTransferFrom(msg.sender, address(this), _amountInMax), "Could not transfer payment token from swapper-payer");
+
+    // Allow the router to spend all this contract's spentToken
+    if (spentToken.allowance(address(this),address(swapRouter)) < _amountInMax) {
+        approveRouterToSpend(address(spentToken));
+    }
+    
+    swapRouter.swapTokensForExactTokens(
+        requestedTotalAmount,
+        _amountInMax,
+        _path,
+        address(this),
+        _deadline
+    );
+    
+    // Allow the payment network to spend all this contract's requestedToken
+    if (requestedToken.allowance(address(this),address(paymentProxy)) < requestedTotalAmount) {
+        approvePaymentProxyToSpend(address(requestedToken));
+    }
+    
+    // Pay the request and fees
+    paymentProxy.transferFromWithReferenceAndFee(
+        address(requestedToken),
+        _to,
+        _amount,
+        _paymentReference,
+        _feeAmount,
+        _feeAddress
+    );
+    
+    // Give the change back to the payer, in both currencies (only spent token should remain)
+
+    if (spentToken.balanceOf(address(this)) > 0) {
+      spentToken.transfer(msg.sender, spentToken.balanceOf(address(this)));
+    }    
+    if (requestedToken.balanceOf(address(this)) > 0) {
+      requestedToken.transfer(msg.sender, requestedToken.balanceOf(address(this)));
```

Comment:
This transfer and the previous one seem to be the riskiest part of the contract.
We need to be 100% sure this will only transfer the funds from the current user/payment.

Reply:
Yes, I thought about that as well. According to me, the contract is never supposed to hold any token. It manipulates 2 per transaction and at the end gives both back. Thus, the only case when the contract would give back "additional" tokens is if someone previously sent tokens to the contract, with a normal erc20.transfer. I don't see good reasons to do that, so I assume it would only be done by mistake. 

Do you see other possible flaws?

Reply:
There is nothing obvious, at least.